File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == ']' and len(stack) and stack[-1] == '[':
                stack.pop()
            elif c == ')' and len(stack) and stack[-1] == '(':
                stack.pop()
            elif c == '}' and len(stack) and stack[-1] == '{':
                stack.pop()
            elif c in {'[','{','('}:
                stack.append(c)
            else:
                return False
        return len(stack) == 0
            
            

        

File: 20-valid-parentheses/valid-parentheses.java
import java.util.Stack;
        
class Solution {
    public boolean isValid(String s) {
        Stack<Character> x = new Stack<>();
        for (int i=0; i<s.length(); i++){
            char c = s.charAt(i);
            if (c == '(' || c == '[' || c == '{') {
                x.push(c);
            } else if (x.isEmpty()) {
                return false; 
            } else if (c == ')' && x.peek() != '(') {
                return false;
            } else if (c == ']' && x.peek() != '[') {
                return false;
            } else if (c == '}' && x.peek() != '{') {
                return false;
            } else {
                x.pop();
            }
        }
        return x.isEmpty();
    }
}